snakemake/executors/flux.py

```diff
@@ -0,0 +1,242 @@
+        """
+        if not flux:
+            sys.exit(
+                "Cannot import flux. Is a cluster available to you with Python bindings?"
+            )
+        self._fexecutor = flux.job.FluxExecutor()
+
+    def cancel(self):
+        """
+        cancel execution, usually by way of control+c. Cleanup is done in
+        shutdown (deleting cached workdirs in Google Cloud Storage
+        """
+        for job in self.active_jobs:
+            if not job.flux_future.done():
+                flux.job.cancel(self.f, job.jobid)
+        self.shutdown()
+
+    def _set_job_resources(self, job):
+        """
+        Given a particular job, generate the resources that it needs,
+        including default regions and the virtual machine configuration
+        """
+        self.default_resources = DefaultResources(
+            from_other=self.workflow.default_resources
+        )
+
+    def get_snakefile(self):
+        assert os.path.exists(self.workflow.main_snakefile)
+        return self.workflow.main_snakefile
+
+    def _get_jobname(self, job):
+        # Use a dummy job name (human readable and also namespaced)
+        return "snakejob-%s-%s-%s" % (self.run_namespace, job.name, job.jobid)
+
+    def format_job_exec(self, job):
+        """
+        We don't want a python -m to run.
+        """
+        prefix = self.get_job_exec_prefix(job)
+        if prefix:
+            prefix += " &&"
+        suffix = self.get_job_exec_suffix(job)
+        if suffix:
+            suffix = f"&& {suffix}"
+        return join_cli_args(
+            [
+                prefix,
+                self.get_envvar_declarations(),
+                "snakemake",
+                format_cli_arg("--snakefile", self.get_snakefile()),
+                self.get_job_args(job),
+                self.general_args,
+                suffix,
+            ]
+        )
+
+    def run(self, job, callback=None, submit_callback=None, error_callback=None):
+        """
+        Submit a job to flux.
+        """
+        super()._run(job)
+
+        from flux.job import JobspecV1
+
+        # This needs to be called after the line above
+        self._prepare_job_formatter()
+
+        # Prepare job resourcces
+        self._set_job_resources(job)
+
+        # The entire snakemake command to run, etc
+        command = self.format_job_exec(job)
+        logger.debug(command)
+
+        # Generate the flux job
+        # flux does not support mem_mb, disk_mb
+        fluxjob = JobspecV1.from_command(command=shlex.split(command))
+
+        # A duration of zero (the default) means unlimited
+        fluxjob.duration = job.resources.get("runtime", 0)
+
+        # Ensure the cwd is the snakemake working directory
+        fluxjob.cwd = self.workdir
+        fluxjob.environment = dict(os.environ)
+        flux_future = self._fexecutor.submit(fluxjob)
+
+        # Waiting for the jobid is a small performance penalty, same as calling flux.job.submit
```

Comment:
I think this is my old comment, meant more for you (in case you cared) than to actually go in snakemake

Reply:
haha but it's useful developer info! I loved your comments. OK to keep? I will definitely remove if that is your preference, or if you'd like me to reword it.